### ISSUE TITLE
Secrets broken on Docker 1.9.1 on non-containerized kubelet

### ISSUE THREAD

**Issue by ghodss** [kubernetes/kubernetes#19622]
I am running Docker 1.9.1 on Kubernetes 1.1.3 on SL6 with a 3.10.0-229.14.1.el7.x86_64 kernel with a NON-containerized kubelet. Secrets mounting works fine on the same host when I am running on Docker 1.7.1.

```
kind: Secret
apiVersion: v1
metadata:
  name:
    mysecret
data:
  abc: ZGVm # "def"
```
```
kind: Pod
apiVersion: v1
metadata:
  name: mypod
spec:
  containers:
    - name: nginx
      image: docker-registry-vip.dev.box.net/jenkins/nginx
      volumeMounts:
        - mountPath: "/secret"
          name: "secret"
          readOnly: true
  volumes:
    - name: secret
      secret:
        secretName: mysecret
```
```
$ kubectl create -f secret.yaml
secret "mysecret" created
$ kubectl create -f pod.yaml
pod "mypod" created
```

All those setup steps are same on Docker 1.7.1 and Docker 1.9.1. This is what happens on Docker 1.7.1 (working):

```
$ # docker exec into the pod's container
$ sudo docker exec -t -i 683 /bin/bash
root@mypod:/# ls -al /secret/
total 8
drwxrwxrwt  2 root root   60 Jan 13 23:18 .
drwxr-xr-x 22 root root 4096 Jan 13 23:26 ..
-r--r--r--  1 root root    3 Jan 13 23:18 abc
root@mypod:/# ls -al /var/run/secrets/kubernetes.io/serviceaccount/
drwxrwxrwt 2 root root   80 Jan 13 23:18 .
drwxr-xr-x 3 root root 4096 Jan 13 23:26 ..
-r--r--r-- 1 root root 7845 Jan 13 23:18 ca.crt
-r--r--r-- 1 root root  846 Jan 13 23:18 token
root@mypod:/#
```

This is what happens when I upgrade the machine to Docker 1.9.1 (not working):

```
$ # docker exec into the pod's container
$ sudo docker exec -t -i 2fc /bin/bash
root@mypod:/# ls -al /secret/
root@mypod:/# ls -al /var/run/secrets/kubernetes.io/serviceaccount/
root@mypod:/#
```

With Docker 1.9.1, the files DO exist on the host:

```
$ sudo ls -al /var/lib/kubelet/pods/ea4ad4dd-ba4b-11e5-bc0f-005056894f70/volumes/kubernetes.io~secret/secret
total 8
drwxrwxrwt 2 root root   60 Jan 13 15:18 .
drwxr-xr-x 4 root root 4096 Jan 13 15:18 ..
-r--r--r-- 1 root root    3 Jan 13 15:18 abc
$ sudo ls -al /var/lib/kubelet/pods/ea4ad4dd-ba4b-11e5-bc0f-005056894f70/volumes/kubernetes.io~secret/default-token-7pgja
total 16
drwxrwxrwt 2 root root   80 Jan 13 15:18 .
drwxr-xr-x 4 root root 4096 Jan 13 15:18 ..
-r--r--r-- 1 root root 7845 Jan 13 15:18 ca.crt
-r--r--r-- 1 root root  846 Jan 13 15:18 token
```

So it seems that for both the kubernetes service account secret and user-specified secrets, the files get pulled down to the host but are not viewable from the container.

**Comment by thockin** [comment]
@pmorie PTAL with priority fo rtriage

**Comment by ghodss** [comment]
@pmorie any update here?

**Comment by pmorie** [comment]
@thockin @ghodss Sorry about that, I didn't see this.  I'll try to repro.

**Comment by pmorie** [comment]
@ghodss Can I see the kubelet logs at `--loglevel=5` ?

**Comment by ghodss** [comment]
I think this is the relevant part https://gist.github.com/anonymous/ed2b85c45330b71a5d98

**Comment by pmorie** [comment]
Nothing obvious from the logs, next step is take a look at the output of `docker inspect` for an affected container.

**Comment by pmorie** [comment]
We were not able to reproduce this on a RHEL 7 system on docker 1.9.  Thinking about next steps for debugging now.

**Comment by pmorie** [comment]
Just a recap of what's been done to debug this:

1.  Verify the secret e2e fails on the affected system
2.  Recreate container bind-mounting a tmpfs from host, we could see files present in the tmpfs mount in the container
3.  Checked the docker metadata, verified the bind-mounts are to/from the correct location
4.  Checked the kubelet log - no problems with the mount.

**Comment by ghodss** [comment]
We've found and fixed the issue. It happened because we switched to using the official Docker init script in our latest Docker RPM build (instead of using the EPEL init script).

For those who are interested, the [official RedHat init script](https://github.com/docker/docker/blob/master/contrib/init/sysvinit-redhat/docker) runs "unshare" before starting up Docker in an attempt to fix an issue where you have something mounted inside a container, and inside that container you create a new mountpoint inside that something which results in that new mountpoint becoming visible outside the container. And from the commit message it seems that in some cases, something outside a container could hold something inside a container open because of that.

However, calling unshare before starting up docker has the effect of hiding all new mounts from the Docker daemon. So when the secrets code creates a tmpfs mount right before importing it into a container, the mount isn't seen by the container.

We've removed the `"$unshare" -m --` line from our init script which seems to have solved the problem.

Thank you for your help @pmorie!
